ppx/browser/ppx_deriving_json_runtime.ml

```diff
@@ -65,6 +70,29 @@ module Of_json = struct
   let option_of_json v_of_json (json : t) =
     if (Obj.magic json : 'a Js.null) == Js.null then None
     else Some (v_of_json json)
+
+  let result_of_json a_of_json b_of_json (json : t) =
```

Comment:
this looks scary but as I understand you did this by using ppx like this and then massaging the code?
```
; echo "
type ('ok, 'err) result = ('ok, 'err) Stdlib.result = Ok of 'ok | Error of 'err [@@deriving json]
" | dune exec ./ppx/browser/ppx_deriving_json_js_test.exe -- -impl -
type ('ok, 'err) result = ('ok, 'err) Stdlib.result =
  | Ok of 'ok
  | Error of 'err [@@deriving json]
include
  struct
    let _ = fun (_ : ('ok, 'err) result) -> ()
    [@@@ocaml.warning "-39-11-27"]
    let rec result_of_json ok_of_json err_of_json =
      (fun x ->
         if Js.Array.isArray x
         then
           let array = (Obj.magic x : Js.Json.t array) in
           let len = Js.Array.length array in
           (if Stdlib.(>) len 0
            then
              let tag = Js.Array.unsafe_get array 0 in
              (if Stdlib.(=) (Js.typeof tag) "string"
               then
                 let tag = (Obj.magic tag : string) in
                 (if Stdlib.(=) tag "Ok"
                  then
                    (if Stdlib.(<>) len 2
                     then
                       Ppx_deriving_json_runtime.of_json_error
                         "expected a JSON array of length 2";
                     Ok (ok_of_json (Js.Array.unsafe_get array 1)))
                  else
                    if Stdlib.(=) tag "Error"
                    then
                      (if Stdlib.(<>) len 2
                       then
                         Ppx_deriving_json_runtime.of_json_error
                           "expected a JSON array of length 2";
                       Error (err_of_json (Js.Array.unsafe_get array 1)))
                    else
                      Ppx_deriving_json_runtime.of_json_error "invalid JSON")
               else
                 Ppx_deriving_json_runtime.of_json_error
                   "expected a non empty JSON array with element being a string")
            else
              Ppx_deriving_json_runtime.of_json_error
                "expected a non empty JSON array")
         else
           Ppx_deriving_json_runtime.of_json_error
             "expected a non empty JSON array" : Js.Json.t ->
                                                   ('ok, 'err) result)
    let _ = result_of_json
    [@@@ocaml.warning "-39-11-27"]
    let rec result_to_json ok_to_json err_to_json =
      (fun x ->
         match x with
         | Ok x_0 ->
             (Obj.magic [|(string_to_json "Ok");(ok_to_json x_0)|] :
             Js.Json.t)
         | Error x_0 ->
             (Obj.magic [|(string_to_json "Error");(err_to_json x_0)|] :
             Js.Json.t) : ('ok, 'err) result -> Js.Json.t)
    let _ = result_to_json
  end[@@ocaml.doc "@inline"][@@merlin.hide ]
```

Reply:
> as I understand you did this by using ppx like this and then massaging the code?

Correct.

Reply:
I just noticed that your example had better naming for the function params. I fixed that directly in `main`.